Clean up some loc assignments

vuonojenmustaturska opened the pull request:
~~Adds per-Z player mob tracking. I've fixed all the cases (even VV was somehow handled) that i could think of where it was possible to make the error pop up, but i very much expect there are some player loc assignments lurking around.~~

~~There will be new, fun, error messages to admins prodding them to contact their nearest coderbus representative for help.~~

~~If this system actually turns out to work alright, it will pave the way to big performance gains in mob handling, and perhaps other optimizations also.~~

~~Probably a very good idea to testmerge this so if there's something fucky, admins won't revolt.~~

scratch all that, doing it in another PR

:cl: Naksu
code: Cleans up some loc assignments
/:cl:

vuonojenmustaturska committed:
Z-level tracking

Supermichael777 commented:
So what does this actually do?

optimumtact commented:
@Supermichael777 keeps a list of players per z level

the underlying goal is to reduce calculation of things not on z levels, for example lavaland mobs can iterate the player list and do dist calculations instead of using view constantly to determine if they need to wake up.

Further enhancements can be stuff like pausing overlay updates to things with no players on the level, disabling atmos etc.

vuonojenmustaturska committed:
remove some leftover debugging

vuonojenmustaturska commented:
I should add, the intent was not to place this in the mobs subsystem but a new global, however when globals are initialized the maps are not loaded and i didn't want to hardcode the number of z-levels and all that. 

When cyberboss delivers on spess management i expect to restore the mobs subsystem to its previous state.

optimumtact commented:
This would make more sense on the subsystem though. Not everything needs to go in the GLOB system, you can think  of the subsystems as namespaced globals anyway

RemieRichards commented on code/controllers/subsystem/mobs.dm:
```diff
@@ -1,14 +1,21 @@
+/datum/controller/subsystem/mobs/Initialize(start_timeofday)
+	by_zlevel.len = world.maxz
+	for (var/i = 1 to world.maxz)
+		by_zlevel.[i] = list()
```
and this compiles because byond is stupid

vuonojenmustaturska replied:
The worst part is that this was the only way I managed to build a variable-length multidimensional list. 

None of the documented ways to initializing a multidimensional list work with variables or work outside of _init() (where maxz is 1 due to maps not being loaded yet), and the usual suspects don't work because insertion functions merge lists so attempting to insert an empty list just does nothing.

At least .len is intentionally read/write and modifying it is documented and supported and it doesn't blow up.

RemieRichards replied:
No no, what you've done here is fine.
I'm talking about ```by_zlevel.[i]```
that's ```.[]``` which isn't an operator, it's two, var lookup and indexing, why it's compiling I haven't a clue besides byond~

vuonojenmustaturska replied:
Oh fuck, that's actually a typo/left-over `.` from .Insert/.Add tests.

I have no idea why it compiles or works either. I'll remove it when i get home.

coiax commented:
Neutral question: Why do we want this?

vuonojenmustaturska commented:
The short answer is that it opens up a lot of avenues for new optimizations down the line, once the lists can be relied upon.

The very first optimization probably being the removal of view() from idle mobs, eliminating roughly 1000 view() calls every 6 seconds. But there are a lot of things that can be done when we have information like "this Z-level has no players", or code cleanups that can happen based on events like "mob just transferred to another Z level"

vuonojenmustaturska committed:
Makes list init cleaner

vuonojenmustaturska commented:
I swear i tried that way of making the empty lists earlier...

vuonojenmustaturska committed:
Update mobs.dm

vuonojenmustaturska committed:
Fix bloodcrawl loc assignment

vuonojenmustaturska committed:
handle Z registration in Destroy()

vuonojenmustaturska committed:
Update fulton.dm

vuonojenmustaturska committed:
Check client in update_z()

optimumtact commented on code/controllers/subsystem/mobs.dm:
```diff
@@ -39,7 +39,7 @@ SUBSYSTEM_DEF(mobs)
 				L.update_z(L.z)
 			else if (!L.client && L.registered_z)
-				message_admins("[L] [ADMIN_FLW(L)] has a Z-registration despite not having a client.")
+				// message_admins("[L] [ADMIN_FLW(L)] has a Z-registration despite not having a client.") removed for spam
```
hehhehe

vuonojenmustaturska committed:
Give up on trying to track Z transitions per case

Turns out mobs can be inside a bag being carried by another mob inside an effect inside a locker being carried by a mecha and trying to handle every scenario like that is just a pain in the ass

vuonojenmustaturska committed:
more cleanup

vuonojenmustaturska committed:
TODO for some other PR

vuonojenmustaturska commented:
I changed the Z tracking error catcher in mobs subsystem to handle all the transitions and gave up on everything else, handling players inside matroskas transiting Z-levels would require shittons of really stupid code.

Cyberboss commented on code/controllers/subsystem/mobs.dm:
```diff
@@ -1,14 +1,19 @@
 	name = "Mobs"
 	priority = 100
-	flags = SS_KEEP_TIMING|SS_NO_INIT
+	flags = SS_KEEP_TIMING
```
Should probably be given an init_order for sanity

Cyberboss commented on code/controllers/subsystem/mobs.dm:
```diff
@@ -27,5 +32,12 @@ SUBSYSTEM_DEF(mobs)
 			M.Life(seconds, times_fired)
 		else
 			GLOB.mob_living_list.Remove(M)
+		if (istype(M, /mob/living)) // This check can be removed if blob cameras are moved into their own system so that they do not require Life() ticks from this subsystem to function
```
We have a global living typecache somewhere

Cyberboss commented on code/controllers/subsystem/mobs.dm:
```diff
@@ -27,5 +32,12 @@ SUBSYSTEM_DEF(mobs)
+			var/turf/T = get_turf(L)
+			if (L.client && L.registered_z != T.z)
+				L.update_z(T.z)
+			else if (!L.client && L.registered_z)
```
cache L.client, T.z, and L.registered_z in proc vars

Cyberboss commented on code/game/objects/effects/step_triggers.dm:
```diff
@@ -132,9 +131,9 @@
-			A.y = rand(teleport_y, teleport_y_offset)
-			A.z = rand(teleport_z, teleport_z_offset)
+			var/turf/T = locate(rand(teleport_x, teleport_x_offset), rand(teleport_y, teleport_y_offset), rand(teleport_z, teleport_z_offset))
+			if (isturf(T))
```
just do if(T) locate always returns a turf

Cyberboss commented on code/modules/mob/living/living_defines.dm:
```diff
@@ -79,3 +79,5 @@
 
 	var/list/obj/effect/proc_holder/abilities = list()
+
+	var/registered_z = null // The Z-level this mob is "registered" to, if it has a client. This is to track non-observing player mobs per z-level.
```
remove the `= null`

vuonojenmustaturska committed:
revert the actual mob tracking part

vuonojenmustaturska committed:
revert continues, requested changes
